ikarus/utils/functionhelper.hh

```diff
@@ -10,35 +10,72 @@
-  static_assert(Concepts::LagrangeNode<std::remove_cvref_t<decltype(localView.tree().child(0))>>,
-                "This function is only supported for Lagrange basis");
-  assert(localView.bound() && "The local view must be bound to an element");
-  const auto& localFE = localView.tree().child(0).finiteElement();
-  lagrangeNodeCoords.resize(localFE.size());
-  std::vector<double> out;
-  for (int i = 0; i < size; i++) {
-    auto ithCoord = [&i](const Dune::FieldVector<double, size>& x) { return x[i]; };
-    localFE.localInterpolation().interpolate(ithCoord, out);
-    for (std::size_t j = 0; j < out.size(); j++)
-      lagrangeNodeCoords[j][i] = out[j];
-  }
-  for (auto& nCoord : lagrangeNodeCoords)
-    nCoord = localView.element().geometry().global(nCoord);
+void obtainLagrangeGlobalNodePositions(const LV& localView,
+                                       std::vector<Dune::FieldVector<double, size>>& lagrangeNodeGlobalCoords) {
+  auto fT = [&]([[maybe_unused]] int nodeNumber, Dune::FieldVector<double, size>&& localCoordinate) {
+    lagrangeNodeGlobalCoords.emplace_back(localView.element().geometry().global(localCoordinate));
+    return false;
+  };
+  forEachLagrangeNodePosition(localView, fT);
+}
+
+/**
+ * \brief A helper function to obtain the global index from the global positions for a Lagrange node
+ * \ingroup utils
+ * \tparam size Size of the global nodal coordinate vector
+ * \tparam Basis Type of the basis.
+ *
+ * \param basis The grid basis.
+ * \param pos Global position
+ * \return Global index
+ */
+template <int size, typename Basis>
+auto globalIndexFromGlobalPosition(const Basis& basis, const Dune::FieldVector<double, size>& pos) {
+  static_assert(Concepts::LagrangeNode<std::remove_cvref_t<decltype(basis.localView().tree().child(0))>>,
+                "globalIndexFromGlobalPosition is only supported for Lagrange power basis");
+  constexpr double tol = 1e-8;
+  using LocalView      = std::remove_cvref_t<decltype(basis.localView())>;
+  using MultiIndex     = typename LocalView::MultiIndex;
+  Dune::HierarchicSearch hSearch(basis.gridView().grid(), basis.gridView().indexSet());
+  const auto& ele = hSearch.findEntity(pos);
+  auto localView  = basis.localView();
+  localView.bind(ele);
+  const auto geo   = localView.element().geometry();
+  const auto& node = localView.tree();
+  std::optional<std::array<MultiIndex, size>> globalIndices;
+
+  auto fT = [&](int nodeNumber, Dune::FieldVector<double, size>&& localCoordinate) {
+    if (Dune::FloatCmp::eq(geo.global(localCoordinate), pos, tol)) {
+      globalIndices.emplace();
+      for (int j = 0; j < size; j++)
+        globalIndices.value()[j] = localView.index(node.child(j).localIndex(nodeNumber));
+      return true;
+    }
+    return false;
+  };
+  forEachLagrangeNodePosition(localView, fT);
+  if (not globalIndices.has_value())
+    DUNE_THROW(Dune::GridError, "No Lagrange node found at the given position in the grid.");
```

Comment:
GridError is not a good error since it is not an error of some of the grid implementations i guess

Reply:
Then I will use `IOError`

Reply:
IOError is for reading and writing from and to files. This makes less sense.